### PR TITLE
Customize LLM config per agent

### PR DESCRIPTION
**Pull request by li-boxuan**
## Intro

Currently, OpenDevin uses a global singleton LLM config and a global singleton agent config. This PR allows customers to configure an LLM config for each agent. A hypothetically useful scenario is to use a cheaper LLM for repo exploration / code search, and a more powerful LLM to actually do the problem solving (CodeActAgent).

Partially solves #2075 (web GUI improvement is not the goal of this PR)

## Breaking changes

Documented in `changelog.md`

## Demo
How the new config.yaml looks like can be found in `config.template.toml` or `test_config.py` in this PR

Task: "Use browser to find out how many stars OpenDevin has on GitHub", default agent: "CodeActAgent", default model: "gpt-4-turbo"

Screenshot 2024-07-06 at 9 50 01 PM

From the backend log we can see:

```
21:48:41 - opendevin:INFO: agent.py:96 - Creating agent CodeActAgent using LLM gpt-4-turbo
21:48:41 - opendevin:INFO: llm.py:137 - Initializing LLM with model: gpt-4-turbo
...
AgentDelegateAction(agent='BrowsingAgent', inputs={'task': 'Sure! Let me browse GitHub to find out how many stars the OpenDevin repository has.. I should start with: Tell me how many stars OpenDevin has on GitHub using google search'}, thought='', action='delegate')
21:49:03 - opendevin:INFO: llm.py:137 - Initializing LLM with model: gpt-4o
21:49:03 - opendevin:INFO: agent_controller.py:239 - [Agent Controller d271d157-ae0b-4420-a55d-d5330d6fd8b6]: start delegate, creating agent BrowsingAgent using LLM LLM(model=gpt-4o)
```

### CODE REVIEW

File: opendevin/core/config.py [hunk 1]
```diff
@@ -162,8 +164,8 @@ class AppConfig(metaclass=Singleton):
         file_uploads_allowed_extensions: List of allowed file extensions for uploads. ['.*'] means all extensions are allowed.
     """
 
-    llm: LLMConfig = field(default_factory=LLMConfig)
-    agent: AgentConfig = field(default_factory=AgentConfig)
+    llms: Dict[str, LLMConfig] = {}
+    agents: Dict[str, AgentConfig] = {}
```
Comment (enyst) [inline opendevin/core/config.py:168]:
```suggestion
    llms: dict[str, LLMConfig] = {}
    agents: dict[str, AgentConfig] = {}
```

Comment (li-boxuan) [reply opendevin/core/config.py:168]: lol you are too fast. Hold on, not ready for review yet 🤣

**Comment by enyst** [comment]
Great to see this! *goes back under a rock*

**Comment by li-boxuan** [comment]
I am still polishing this PR but it's basically ready for review.

File: config.template.toml [hunk 2]
```diff
@@ -159,8 +172,12 @@ model = "gpt-4o"
...
 #memory_max_threads = 2
 
-# Name of the agent
-#name = "CodeActAgent"
+# LLM config group to use
+#llm_config = 'llm'
+
+[BrowsingAgent]
+# e.g. use a cheaper model for BrowsingAgent
+llm_config = 'gpt3_llm'
```
Comment (neubig) [inline config.template.toml:180]: I don't think `BrowsingAgent` would work as well with gpt-3. I know this is supposed to be an example, but some people might take this as a recommendation, so perhaps we could keep it with gpt-4 by default.

File: opendevin/controller/agent_controller.py [hunk 3]
```diff
@@ -76,6 +76,7 @@ def __init__(
         """
         self._step_lock = asyncio.Lock()
         self.id = sid
+        print('\n\n\n\\ agent =', agent)
```
Comment (neubig) [inline opendevin/controller/agent_controller.py:79]: Is this intentional? Maybe it should be a logging command or removed?

File: opendevin/core/config.py [hunk 4]
```diff
@@ -338,6 +373,11 @@ def get_optional_type(union_type: UnionType) -> Any:
     # helper function to set attributes based on env vars
     def set_attr_from_env(sub_config: Any, prefix=''):
         """Set attributes of a config dataclass based on environment variables."""
+        print(
```
Comment (neubig) [inline opendevin/core/config.py:376]: Logging or delete

**Review by neubig — APPROVED** [review]
Basically LGTM, thanks for doing this! Just a few comments about printing/logging, but happy to merge this in after that.

File: opendevin/controller/agent_controller.py (hunk 3 above)
Comment (li-boxuan) [reply opendevin/controller/agent_controller.py:79]: This is my bad habit of using printing as debugging statements :) Will remove

File: config.template.toml [hunk 5]
```diff
@@ -79,8 +79,12 @@ persist_sandbox = false
 # Use host network
 #use_host_network = false
 
+# Name of the default agent
+#default_agent = "CodeActAgent"
+
 #################################### LLM #####################################
-# Configuration for the LLM model
+# Configuration for LLM models (group name ends with 'llm')
```
Comment (enyst) [inline config.template.toml:86]: Does it end with 'llm'?

File: evaluation/agent_bench/README.md [hunk 6]
```diff
@@ -29,12 +29,12 @@ enable_auto_lint = true
...
 timeout = 120
 
-[eval_gpt35_turbo]
+[llm.eval_gpt35_turbo]
 model = "gpt-3.5-turbo"
 api_key = "sk-123"
 temperature = 0.0
 
-[eval_gpt4o]
+[llm.eval_gpt4o]
```
Comment (enyst) [inline evaluation/agent_bench/README.md:37]: Wow this is lovely. (it starts with 'llm' rather than end)

File: opendevin/core/utils/singleton.py [hunk 7]
```diff
@@ -21,8 +21,10 @@ def reset(cls):
...
-                if dataclasses.is_dataclass(field.type):
-                    setattr(instance, field.name, field.type())
+            for field_info in dataclasses.fields(instance_type):
+                if dataclasses.is_dataclass(field_info.type):
+                    setattr(instance, field_info.name, field_info.type())
+                elif field_info.default_factory is not dataclasses.MISSING:
+                    setattr(instance, field_info.name, field_info.default_factory())
                 else:
-                    setattr(instance, field.name, field.default)
+                    setattr(instance, field_info.name, field_info.default)
```
Comment (enyst) [inline opendevin/core/utils/singleton.py:30]: Thanks! (and here we see why copilot ain't that good...)

File: opendevin/llm/bedrock.py [hunk 8]
```diff
@@ -5,9 +5,9 @@
 from opendevin.core.config import config
 from opendevin.core.logger import opendevin_logger as logger
 
-AWS_ACCESS_KEY_ID = config.llm.aws_access_key_id
-AWS_SECRET_ACCESS_KEY = config.llm.aws_secret_access_key
-AWS_REGION_NAME = config.llm.aws_region_name
+AWS_ACCESS_KEY_ID = config.get_llm_config().aws_access_key_id
+AWS_SECRET_ACCESS_KEY = config.get_llm_config().aws_secret_access_key
+AWS_REGION_NAME = config.get_llm_config().aws_region_name
```
Comment (enyst) [inline opendevin/llm/bedrock.py:10]: Will it work with bedrock if bedrock vars are set in some llm config, but not default?

File: opendevin/core/config.py [hunk 9]
```diff
@@ -243,6 +245,39 @@ class AppConfig(metaclass=Singleton):
...
+            self.agents['agent'] = AgentConfig()
+        return self.agents['agent']
+
+    def set_agent_config(self, value: AgentConfig, name='agent'):
+        self.agents[name] = value
+
+    def get_llm_config_from_agent(self, name='agent') -> LLMConfig:
+        agent_config: AgentConfig = self.get_agent_config(name)
+        llm_config_name = agent_config.llm_config
+        return self.get_llm_config(llm_config_name)
```
Comment (enyst) [inline opendevin/core/config.py:279]: This reads nicely. Hey about time there's some nice code around here. ❤️

File: opendevin/core/utils/singleton.py (hunk 7 above)
Comment (li-boxuan) [reply opendevin/core/utils/singleton.py:30]: this piece of code was generated by GPT too 😆

File: config.template.toml (hunk 5 above)
Comment (li-boxuan) [reply config.template.toml:86]: Ah thx, I went back and forth on design decisions and this was a leftover I forgot to clean up

File: opendevin/llm/bedrock.py (hunk 8 above)
Comment (li-boxuan) [reply opendevin/llm/bedrock.py:10]: It only works if it's set as default model. This code was a hack IMO and I don't know how to make it better.

Comment (enyst) [reply opendevin/llm/bedrock.py:10]: We can add some TODO here. Aws-specific stuff is a bit oddly placed on top of our existing vars, and I seem to recall we also want to support this way a few other widely used providers. We may need to think about a better solution how to plug them in.

(also, this code was added when litellm was bugged IIRC, so now users have another option to make it work, although still as default)

File: opendevin/core/main.py [hunk 10]
```diff
@@ -163,7 +163,7 @@ async def on_event(event: Event):
             raise ValueError(f'Invalid toml file, cannot read {args.llm_config}')
         llm = LLM(llm_config=llm_config)
     else:
-        llm = LLM(model=args.model_name)
+        llm = LLM(llm_config=config.get_llm_config_from_agent(args.agent_cls))
```
Comment (li-boxuan) [inline opendevin/core/main.py:166]: This breaks `--model-name` command line argument.

Looks like we should just remove `--model-name` from CLI.

Comment (tobitege) [reply opendevin/core/main.py:166]: Just fyi, did a cursory search: while `model-name` is barely found (just for processing), the shorthand `-m` is used in a couple of .py/.md files, also evaluations.

Comment (enyst) [reply opendevin/core/main.py:166]: I would agree here, been thinking about removing it before. It's a bit unexpected now too, it's the most specific but llm_config overrides it.

It affects IDE configurations and cli personal scripts. We may want to give a heads-up on slack.

**Comment by enyst** [comment]
I didn't get to play with it yet, and it's not fully clear to me, about breaking changes, what exactly are the old options that no longer work. llm_configs with random names, right? 

I'd appreciate a list of not working stuff 🤔, because I saw you added compatibility code, and it seems to me that older compatibility code will also continue to work.

File: opendevin/core/main.py (hunk 10 above)
Comment (li-boxuan) [reply opendevin/core/main.py:166]: IMO `main` is okay to break without notice. Will give a heads-up on slack without @ everyone.

Release is trickier - we need to have a release note. @mamoodi Do you have an idea where we might document all breaking changes? A shared doc would suffice. Alternatively, we could add something like a `changelog.md` to document that.

Comment (li-boxuan) [reply opendevin/core/main.py:166]: I am going to go ahead and create a changelog.md doc.

File: docs/modules/usage/changelog.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+## 0.8 (release date: ??)
+
+### Config breaking changes
+
+In this release we introduced a few breaking changes to backend configurations.
+If you have only been using OpenDevin via frontend (web GUI), nothing needs
+to be taken care of.
+
+Here's a list of breaking changes in configs. They only apply to customers who
```
Comment (enyst) [inline docs/modules/usage/changelog.md:15]: s/customers/users

**Review by enyst — APPROVED** [review]

**Comment by enyst** [comment]
Thanks for adding documentation on the breaking changes. I admit I'm heavily in favor of keeping compatibility as much as possible here, but this looks very good to me: most continue to be compatible, docker app not affected, and I think the couple of cli vars remaining will be okay in exchange. This PR is adding a flexibility that was long due and long requested.

Thank you for this. ❤️

**Comment by tobitege** [comment]
Awesome job! Long overdue, brings a lot of flexibility! Thanks! 🏅

File: docs/modules/usage/changelog.md (hunk 11 above)
Comment (li-boxuan) [inline docs/modules/usage/changelog.md:15]:
```suggestion
Here's a list of breaking changes in configs. They only apply to users who
```

**Pull request merged**